docs/digits_diffusion_model.md

```diff
@@ -0,0 +1,822 @@
+)
+
+# Optimizer configuration
+optimizer = nnx.Optimizer(model, optax.chain(
+    optax.clip_by_global_norm(0.5),
+    optax.adamw(
+        learning_rate=schedule_fn,
+        weight_decay=2e-5,
+        b1=0.9,
+        b2=0.999,
+        eps=1e-8
+    )
+))
+
+# Model initialization with dummy input
+dummy_input = jnp.ones((1, 8, 8, 1))
+dummy_t = jnp.zeros((1,), dtype=jnp.int32)
+output = model(dummy_input, dummy_t)
+
+print("Input shape:", dummy_input.shape)
+print("Output shape:", output.shape)
+print("\nModel initialized successfully")
+```
+
++++ {"id": "LrzTfkDPJm2X"}
+
+## Training Loop Implementation
+
+Here you have the main training loop for the diffusion model with:
+- Progressive timestep sampling strategy
+- [Exponential Moving Average (EMA)](https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average) loss tracking
+- Adaptive noise generation
+
+```{code-cell}
+---
+colab:
+  base_uri: https://localhost:8080/
+id: ZnQqHCAoVfi1
+outputId: a105e2de-ba88-44d0-bad5-3a9a69e54826
+---
+# Initialize training metrics
+losses: List[float] = []                # Store EMA loss history
+moving_avg_loss: Optional[float] = None  # EMA of the loss value
+beta: float = 0.99                      # EMA decay factor for loss smoothing
+
+for epoch in range(num_epochs + 1):
+    # Split PRNG key for independent random operations
+    key, subkey1 = jax.random.split(key)
+    key, subkey2 = jax.random.split(key)
+
+    # Progressive timestep sampling - weights early steps more heavily as training progresses
+    # This helps model focus on fine details in later epochs while maintaining stability
+    progress = epoch / num_epochs
+    t_weights = jnp.linspace(1.0, 0.1 * (1.0 - progress), num_steps)
+    t = jax.random.choice(
+        subkey1,
+        num_steps,
+        shape=(images_train.shape[0],),
+        p=t_weights/t_weights.sum()
+    )
+
+    # Generate Gaussian noise for current batch
+    noise = jax.random.normal(subkey2, images_train.shape)
+
+    # Execute training step with noise prediction and parameter updates
+    loss = train_step(
+        model, optimizer, images_train, t, noise,
+        diffusion.sqrt_alpha_cumulative, diffusion.sqrt_one_minus_alpha_cumulative
+    )
+
+    # Update exponential moving average of loss for smoother tracking
+    if moving_avg_loss is None:
+        moving_avg_loss = loss
+    else:
+        moving_avg_loss = beta * moving_avg_loss + (1 - beta) * loss
+
+    losses.append(moving_avg_loss)
+
+    # Log training progress at regular intervals
+    if epoch % 100 == 0:
+        print(f"Epoch {epoch}, Loss: {moving_avg_loss:.4f}")
+
+print("\nTraining completed.")
+```
+
++++ {"id": "SHX69WgxKUs7"}
+
+
+
++++ {"id": "s-iqch4HKlBV"}
+
+## Training loss visualization
+
+It uses a logarithmic scale to better display the exponential decay of the loss values over time. This representation helps identify both early rapid improvements and later fine-tuning phases of the training process.
```

Comment:
```suggestion
We use a logarithmic scale to better display the exponential decay of the loss values over time. This representation helps identify both early rapid improvements and later fine-tuning phases of the training process.
```

Reply:
As in the common Google tech writing guideline, I do prefer to keep it without the "We language".